### ISSUE TITLE
Manuall order and shipping method instance ID

### ISSUE THREAD

**Issue by vyskoczilova** [woothemes/woocommerce#38481]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When I create order manually, I can retrieve `method_id` but not `instance_id` (it returns 0) although it should have been parsed upon the address added by user. 

### Expected behavior

Return instance_id based on the address (billing or shiping if added)

### Actual behavior

Returns `0`

### Steps to reproduce

1. Create an manuall order with billing address, one product and one existing shipping (not N/A or Other).
2. Use this code:

```
$order = wc_get_order($order_id);
$shipping_methods = $order->get_items('shipping');
$shipping_method = reset($shipping_methods);
var_dump($shipping_method->get_instance_id()); // 0
var_dump($shipping_method->get_method_id()); // correct method ID
```

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): https://wc-mailstep.ddev.site
Site address (URL): https://wc-mailstep.ddev.site
WC Version: 7.7.0
REST API Version: ✔ 7.7.0
WC Blocks Version: ✔ 10.0.4
Action Scheduler Version: ✔ 3.5.4
Log Directory Writable: ✔
WP Version: 6.2
WP Multisite: –
WP Memory Limit: 1 GB
WP Debug Mode: ✔
WP Cron: ✔
Language: cs_CZ
External object cache: –

### Server Environment ###

Server Info: nginx/1.22.1
PHP Version: 8.0.26
PHP Post Max Size: 100 MB
PHP Time Limit: 600
PHP Max Input Vars: 5000
cURL Version: 7.74.0
OpenSSL/1.1.1n

SUHOSIN Installed: –
MySQL Version: 10.4.27-MariaDB-1:10.4.27+maria~ubu2004-log
Max Upload Size: 100 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.4.1
WC Database Prefix: wp_
Total Database Size: 5.80MB
Database Data Size: 4.24MB
Database Index Size: 1.56MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 3.06MB + Index: 0.06MB + Engine InnoDB
wp_packetery_carrier: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_packetery_log: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_packetery_order: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_postmeta: Data: 0.11MB + Index: 0.09MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_snippets: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 22
page: 6
post: 1
product: 19
product_variation: 6
revision: 2
shop_order: 5
wp_global_styles: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (6) ###

Query Monitor: by John Blackbourn – 3.12.3
Code Snippets: by Code Snippets Pro – 3.4.0
Kybernaut Mailstep: by Karolína Vyskočilová – 1.7.2-beta
Mailstep Debug: by Karolína Vyskočilová – 1.0
Packeta: by Zásilkovna s.r.o. – 1.5.4
WooCommerce: by Automattic – 7.7.0

### Inactive Plugins (3) ###

WooCommerce Sequential Order Numbers Pro: by SkyVerge – 1.18.1
WooCommerce Stripe Gateway: by WooCommerce – 7.4.0
WPify Woo: by WPify s.r.o. – 3.9.12

### Dropin Plugins (1) ###

db.php: Query Monitor Database Class (Drop-in)

### Settings ###

API Enabled: –
Force SSL: –
Currency: CZK (Kč)
Currency Position: right_space
Thousand Separator: 
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: –
HPOS feature enabled: –
Order datastore: WC_Order_Data_Store_CPT
HPOS data sync enabled: –

### WC Pages ###

Shop base: #63 - /obchod/
Cart: #64 - /kosik/
Checkout: #65 - /pokladna/
My account: #66 - /muj-ucet/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 4.2.0 (update to version 4.3.0 is available)
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: product-block-editor
minified-js
new-product-management-experience
product-variation-management
settings

Daily Cron: ✔ Next scheduled: 2023-05-26 17:09:51 +00:00
Options: ✔
Notes: 76
Onboarding: completed

### Action Scheduler ###

Complete: 28
Oldest: 2023-04-26 12:55:47 +0000
Newest: 2023-05-26 11:58:43 +0000

Failed: 4
Oldest: 2023-03-07 11:32:06 +0000
Newest: 2023-05-25 11:21:16 +0000

Pending: 2
Oldest: 2023-05-27 11:58:43 +0000
Newest: 2023-05-27 11:58:43 +0000

### Status report information ###

Generated at: 2023-05-26 13:13:59 +00:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by flakas** [comment]
Hi @vyskoczilova - thanks for the suggestion.

I know this has been open for a very long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I've reviewed this from a product perspective, and the `instance_id` seems to be `0` for manually created orders because the manually added shipping line isn't actually an instance, and it's not always possible to automatically resolve one: multiple zones could match the method, the admin may have used custom rates, address matches no zones, method is "Other"/free text.

If this improvement is still important to you, could you please share more about your use case, how you're using the API and how would `instance_id` being available for manual orders help you?

I'm going to close this as not planned for the time being. But if the context changes or there's a stronger use case we should consider, we can revisit it.

**Comment by vyskoczilova** [comment]
This is important for me, since I have a drop shipping integration [Kybernaut Mailstep](https://kybernaut.cz/en/plugins/kybernaut-mailstep/) that needs to resolve on the instance ID, and there is no way out of it. I resolved it for my case this way (when it returns `0`):

```php
/**
 * Get Shipping Instance ID
 * @param string $method_id Method ID.
 * @param WC_Order $order Order.
 * @return string
 */
public static function getShippingInstanceIdFromZonesByCountry($method_id, $order)
{

	// Get the country.
	$country = $order->get_shipping_country() ? $order->get_shipping_country() : $order->get_billing_country();

	// Get all shipping zones
	$shipping_zones = \WC_Shipping_Zones::get_zones();

	// Iterate through each shipping zone
	foreach ($shipping_zones as $shipping_zone) {
		
		$zone = new \WC_Shipping_Zone($shipping_zone["id"]);

		// Get all locations within the zone
		$zone_locations = $zone->get_zone_locations();

		// Check if the zone supports shipping to the selected country
		foreach ($zone_locations as $location) {
			if ($location->code === $country && $location->type === 'country') {

				// Get all shipping methods within the zone.
				$shipping_methods = $zone->get_shipping_methods();

				// Iterate through each shipping method.
				foreach ($shipping_methods as $shipping_method) {

					// Check if the shipping method is the one we're looking for.
					if ($shipping_method->id === $method_id) {
						return $shipping_method->get_instance_id();
						break;
					}
				}

				break 2; // No need to check other zones if the country is found in one zone
			}
		}

		return 0;
	}
}
```

**Comment by flakas** [comment]
@vyskoczilova I appreciate this is important for your extension, and thank you for sharing the code snippet!
Glad to hear you were able to find a workaround for the time being!

The code snippet is focused on matching countries. Do you ever find a need to match based on state/postcode too, or if your users can have multiple instances match (e.g. Flat rate Standard or Flat rate Express)?

I'm reopening the ticket and will ask the engineering team for a perspective.

I can see several potential options:
A) We try to do instance matching where possible, making sure to cover as many cases as possible, but falling back to the current `instance_id=0` value where a match is not possible (e.g. In `Orders > New > Add item > Add shipping > user chooses N/A or Other`, or multiple shipping method instances match), without UI changes.
Could potentially work since it appears in many cases methods are unique with only once instance in them.

B) We change the UI how shipping is added to a manually created order, and instead of merchants choosing configured+unconfigured methods users would need to choose among configured instances. That way shipping method instances are explicitly chosen by the user, and we handle the case of multiple instances matching.

**Comment by vyskoczilova** [comment]
Hi @flakas, yeah, it works for couple of years like this (since the time I opened the issue), I'm based in Europe and my clients too, so we don't generally use too much post code matching (only for local city deliveries, but in that case, you won't use the drop-shipping company I built my plugin for), so I didn't needed that myself.

I think A) is the way I did, but what I see, B) is what users would love too in many cases, they would love to pick existing shipping method instead of reinventing it every time somebody call them or they create an order inside their office while talking with the client.

Keep me posted and if I can help, I'm happy to do so.

Best regards,
Karolína
